### ISSUE TITLE
Prevent from adding the same record on build when using source record in attribute within belongs_to on using has_many_inversing

### ISSUE THREAD

**Issue by dmitry** [rails/rails#40378]
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile do
  source 'https://rubygems.org'

  ruby '2.6.1'

  gem 'activesupport', path: '../../activesupport/'
  gem 'activemodel', path: '../../activemodel/'
  gem 'activerecord', path: '../'
  gem 'sqlite3'
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Base.has_many_inversing = true

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
  end

  create_table :phone_numbers, force: true do |t|
    t.references :user, null: false
  end
end

class User < ActiveRecord::Base
  has_many :phone_numbers, inverse_of: :user
end

class PhoneNumber < ActiveRecord::Base
  belongs_to :user, inverse_of: :phone_numbers
end

class BugTest < Minitest::Test
  def test_build_with_belongs_to_association_record_in_attribute
    u = User.create
    u.phone_numbers.build(
      user: u
    )
    assert_equal 1, u.phone_numbers.size
    u.save!
    assert_equal 1, u.phone_numbers.size
  end
end

```

### Expected behavior

When using .build on has_many association and passing source (belongs_to) association as attributes, it should store and return only one associated record.

### Actual behavior

It returns 2 records, because `add_to_target` not replaces already existing record, but adds additional one. In result there are two identical records in has_many association.

### System configuration
**Rails version**: 6.1.alpha
**Ruby version**: 2.7

**Comment by gmcgibbon** [comment]
Similar to https://github.com/rails/rails/issues/38312, but on a different association end. To be clear, the two objects in `u.phone_numbers` are identical in that they have the same object ID. This is not a regression.

**Comment by jonathanhefner** [comment]
Fixed by #40379.
